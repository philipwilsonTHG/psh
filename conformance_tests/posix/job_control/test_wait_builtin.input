# Test wait builtin functionality
# PSH supports wait builtin for process synchronization (v0.57.4 feature)

# Test wait with specific PID
echo "Testing wait with specific PID:"
true &
sleep_pid=$!
echo "Started background job"
wait $sleep_pid
echo "Wait for specific PID completed, status: $?"

# Test wait with multiple PIDs
echo "Testing wait with multiple PIDs:"
true &
pid1=$!
true &
pid2=$!
echo "Started two background jobs"
wait $pid1 $pid2
echo "Wait for multiple PIDs completed, status: $?"

# Test wait with no arguments (wait for all)
echo "Testing wait with no arguments:"
true &
true &
true &
echo "Started 3 background jobs"
wait
echo "Wait for all jobs completed"

# Test wait with non-existent PID
echo "Testing wait with non-existent PID:"
wait 99999 2>/dev/null
echo "Wait for non-existent PID completed, status: $?"

# Test wait with job that fails
echo "Testing wait with failing job:"
false &
false_pid=$!
echo "Started failing job"
wait $false_pid
echo "Wait for failing job completed, status: $?"

# Test wait in pipeline context
echo "Testing wait in pipeline:"
(echo "pipeline test" | cat) &
pipe_pid=$!
wait $pipe_pid
echo "Pipeline wait completed"

# Test wait with command substitution
echo "Testing wait with command substitution background job:"
(echo "command substitution background") &
cmd_sub_pid=$!
result=$(wait $cmd_sub_pid; echo "wait completed")
echo "Result: $result"

# Test wait behavior when no background jobs exist
echo "Testing wait when no background jobs exist:"
wait 2>/dev/null
echo "Wait with no jobs completed, status: $?"

# Test sequential waits
echo "Testing sequential waits:"
true &
first_pid=$!
true &
second_pid=$!

echo "Waiting for first job..."
wait $first_pid
echo "First job completed, status: $?"

echo "Waiting for second job..."
wait $second_pid
echo "Second job completed, status: $?"

echo "Wait builtin test completed"