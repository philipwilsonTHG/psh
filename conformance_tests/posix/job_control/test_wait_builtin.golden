Testing wait with specific PID:
Started background job
Wait for specific PID completed, status: 127
Testing wait with multiple PIDs:
Started two background jobs
Wait for multiple PIDs completed, status: 127
Testing wait with no arguments:
Started 3 background jobs
Wait for all jobs completed
Testing wait with non-existent PID:
Wait for non-existent PID completed, status: 127
Testing wait with failing job:
Started failing job
Wait for failing job completed, status: 127
Testing wait in pipeline:
Pipeline wait completed
Testing wait with command substitution background job:
Result: wait completed
Testing wait when no background jobs exist:
Wait with no jobs completed, status: 0
Testing sequential waits:
Waiting for first job...
First job completed, status: 127
Waiting for second job...
Second job completed, status: 127
Wait builtin test completed
